juju/juju
# Initial work to allow juju trust for sidecar charms

Add support for `juju trust` with sidecar charms.
This needed extra work to create, for each deployed application, a service account, roles, and rolebindings. The application pod is then given this service account rather than the cluster default. The initial role policy rules are the same as those used for pod spec charms. 

Still need to implement this for pod spec charms

Also, a bigger issue is that the unit agent and k8s application provisioner workers both watch for trust changes separately. So the uniter agent might fire the config-change hook to signal a trust change before the provisioner has updated the service account roles. This will need to be addressed.

## QA steps

```
juju bootstrap microk8s --config logging-config="juju.kubernetes.provider.application=DEBUG"
juju deploy facundo-snappass-test
```

Check application pod has service account set
```
kubectl -n controller-ian get -o json pod/facundo-snappass-test-0 | jq .spec.serviceAccount
facundo-snappass-test
```

Check app service account, role bindings, roles are created.
```
kubectl -n controller-ian get -o json serviceaccount/facundo-snappass-test
kubectl -n controller-ian get -o json roles/facundo-snappass-test
kubectl -n controller-ian get -o json rolebindings/facundo-snappass-test
```

Trust the app
```
juju trust facundo-snappass-test
```
Debug logs should have
```
DEBUG juju.kubernetes.provider.application Application "facundo-snappass-test", trust true
```

Role should have new rules:
```
kubectl -n controller-ian get -o json roles/facundo-snappass-test : jq .rules
[
    {
        "apiGroups": [
            "*"
        ],
        "resources": [
            "*"
        ],
        "verbs": [
            "*"
        ]
    }
]

```

Remove trust:

```
juju trust facundo-snappass-test --remove
kubectl -n controller-ian get -o json roles/facundo-snappass-test | jq .rules
[
  {
    "apiGroups": [
      ""
    ],
    "resources": [
      "pods",
      "services"
    ],
    "verbs": [
      "get",
      "list",
      "patch"
    ]
  },
  {
    "apiGroups": [
      ""
    ],
    "resources": [
      "pods/exec"
    ],
    "verbs": [
      "create"
    ]
  }
]
```
